Made clock triggers part of the main tutorial path (#694)
cylc/cylc-doc: src/tutorial/scheduling/datetime-cycling.rst

* ``baz[^]``: the task ``baz`` from the initial cycle point.

.. _tutorial-clock-triggers:

Clock Triggers
--------------

.. ifnotslides::

   In Cylc, :term:`cycle points <cycle point>` are just task labels. Tasks are
   triggered when their dependencies are met, regardless of cycle point.
   But *clock triggers* can be used to force tasks to wait for a particular
   real time, relative to their cycle point, before running.
   This is necessary for certain operational and monitoring systems, e.g. for
   tasks that process real-time data.

>    For example in the following workflow the tasks at cycle
>    ``2050-01-01T12Z`` will wait
>    until 12:00 on the 1st of January 2050 before running:

Reviewer: ```suggestion
   For example, in the following workflow, the task ``do_this_on_or_after_noon``
   will wait until 12:00 every day before running (starting on January 1st 2050):
```

* ``baz[^]``: the task ``baz`` from the initial cycle point.

.. _tutorial-clock-triggers:

Clock Triggers
--------------

.. ifnotslides::

   In Cylc, :term:`cycle points <cycle point>` are just task labels. Tasks are
   triggered when their dependencies are met, regardless of cycle point.
   But *clock triggers* can be used to force tasks to wait for a particular
   real time, relative to their cycle point, before running.
   This is necessary for certain operational and monitoring systems, e.g. for
   tasks that process real-time data.

   For example in the following workflow the cycle ``2050-01-01T12Z`` will wait

>    until 12:00 on the 1st of January 2050 before running:

Reviewer 2: Note, the cycle doesn't wait, the task does.

This isn't an important distinction in this example (only one task) but is where multiple parentless tasks are present, such as the tutorial workflow.

Author: ```suggestion
   For example in the following workflow the tasks at cycle 
   ``2050-01-01T12Z`` will wait
   until 12:00 on the 1st of January 2050 before running:
```

Reviewer: ```suggestion
   For example, in the following workflow, the task ``do_this_on_or_after_noon``
   will wait until 12:00 every day before running (starting on January 1st 2050):
```

.. code-block:: cylc

   [scheduling]
       initial cycle point = 2050-01-01T00Z
       [[xtriggers]]

>            on_time = wall_clock()
>        [[graph]]
>            # "daily" will run, at the earliest, one hour before midday.
>            T12 = @on_time => do_this_on_or_after_noon

Reviewer: - I think `clock` rather than `on_time` would be a more conventional name
- Comment doesn't match?
```suggestion
           clock = wall_clock()
       [[graph]]
           T12 = @clock => do_this_on_or_after_noon
```

>        [[xtriggers]]
>            clock = wall_clock()
>        [[graph]]
>            T12 = @clock => do_this_on_or_after_noon
> 
> .. ifnotslides::
> 
>    Clock triggers take an argument - ``offset`` which allows you to specify
>    an offset. For example, it might be ok for a task to run up to an hour
>    before the specified cycle point:

Reviewer 2: Suggest trimming out xtrigger detail and keeping it to the bare minimum for the tutorial:

```suggestion
       [[graph]]
           T12 = @wall_clock => do_this_on_or_after_noon
```

>        [[xtriggers]]
>            clock = wall_clock()
>        [[graph]]
>            T12 = @clock => do_this_on_or_after_noon
> 
> .. ifnotslides::
> 
>    Clock triggers take an argument - ``offset`` which allows you to specify
>    an offset. For example, it might be ok for a task to run up to an hour
>    before the specified cycle point:
> 
> ``wall_clock(-PT1H)``

Reviewer 2: Suggest trimming out xtrigger detail and keeping it to the bare minimum for the tutorial:

```suggestion
       [[graph]]
           T12 = @wall_clock => do_this_on_or_after_noon
```

.. ifnotslides::

   All xtriggers can be suffixed with ``:interval`` indicating how often the

>    scheduler will run the check. The default is 10 seconds.

Reviewer 2: All xtriggers, other than wall_clock that is!

>    All xtriggers take a ``:recurrance`` indicating how often the scheduler
>    will run the check.
> 
>    It is good practice not to check more often than necessary:
> 
> ``wall_clock():PT17M``

Reviewer: ```suggestion
   All xtriggers can be suffixed with ``:interval`` indicating how often the
   scheduler will run the check. The default is 10 seconds.
   
   It is good practice not to check more often than necessary. This will
   depend on the specific use case. For example:

``wall_clock():PT20M``
```
(17 minutes is a weird interval that is triggering a sort of OCD response in me 😄)

> ``wall_clock():PT20M``

Reviewer 2: I don't think providing an interval for wall_clock does anything because it uses a special implementation rather than actually running a Python script via the subprocpool.

Reviewer: Ah right, good point

Author: I had been going to question this based on the first comment: A noddy test shows that you are correct.

> .. ifnotslides::
> 
>    All xtriggers can be suffixed with ``:interval`` indicating how often the
>    scheduler will run the check. The default is 10 seconds.
>    
>    It is good practice not to check more often than necessary. This will
>    depend on the specific use case. For example:
> 
> ``wall_clock():PT20M``

Reviewer 2: This is a bit too detailed for the tutorial (we haven't taught them to write their own xtriggers yet):

```suggestion
```

      To ensure that the ``get_observations_<location>`` tasks run only
      after the time of the observation, add a clock trigger.
      Observations will be available by 10 minutes past the hour.
      By default, the scheduler will check the clock trigger every 10
      seconds, but there is no point in doing this for any interval
      less than around 5 minutes in this case.

      .. hint::

         See :ref:`tutorial-clock-triggers`

      .. spoiler:: Solution warning

         .. note::

            There are a number of ways of writing the dependency.
            You can be correct and not look _exactly_ like this.

         .. code-block:: diff

           [scheduling]
               initial cycle point = 20000101T00Z
           +   [[xtriggers]]
           +       obs_clock_trigger = wall_clock(PT10M):PT5M
               [[graph]]
                   PT3H = """
           +           @obs_clock_trigger => get_observations_aldergrove => consolidate_observations
           +           @obs_clock_trigger => get_observations_heathrow => consolidate_observations
           +           @obs_clock_trigger => get_observations_aldergrove => consolidate_observations

>            +           @obs_clock_trigger => get_observations_camborne => consolidate_observations

Reviewer: Diff is missing the removals

>            +           @obs_clock_trigger => get_observations_aldergrove => consolidate_observations
>            +           @obs_clock_trigger => get_observations_heathrow => consolidate_observations
>            +           @obs_clock_trigger => get_observations_aldergrove => consolidate_observations
>            +           @obs_clock_trigger => get_observations_camborne => consolidate_observations

Author: ```suggestion
           -            get_observations_aldergrove => consolidate_observations
           +           @obs_clock_trigger => get_observations_aldergrove => consolidate_observations
           -            get_observations_heathrow => consolidate_observations
           +           @obs_clock_trigger => get_observations_heathrow => consolidate_observations
           -            get_observations_aldergrove => consolidate_observations
           +           @obs_clock_trigger => get_observations_aldergrove => consolidate_observations
           -            get_observations_camborne => consolidate_observations
           +           @obs_clock_trigger => get_observations_camborne => consolidate_observations
```

         +          +PT12H/PT6H = """
         +              forecast[-PT6H] => forecast
         +          """

   #. **Clock Triggers**

      To ensure that the ``get_observations_<location>`` tasks run only

>       after the time of the observation add a clock trigger.
>       Observations are available by 10 minutes past the hour.
>       We don't want to overburden the Cylc server, so check for observations
>       every 5 minutes.

Reviewer: Overburden is perhaps too strong a word for this example with 1 xtrigger?
```suggestion
      after the time of the observation, add a clock trigger.
      Observations will be available by 10 minutes past the hour.
      By default, the scheduler will check the clock trigger every 10
      seconds, but there is no point in doing this for any interval
      less than around 5 minutes in this case.
```

>       .. spoiler:: Hint hint
> 
>          See :ref:`tutorial-clock-triggers`

Reviewer: ```suggestion
      .. hint::

         See :ref:`tutorial-clock-triggers`
```

      .. spoiler:: Solution warning

         .. note::

            There are a number of ways of writing the dependency: This

>             example shows multiple patterns.

Reviewer: It's fine to mention there are multiple ways of doing this, but I would only show 1 way for consistency.

         +          +PT12H/PT6H = """
         +              forecast[-PT6H] => forecast
         +          """

   #. **Clock Triggers**

      To ensure that the ``get_observations_<location>`` tasks run only
      after the time of the observation, add a clock trigger.
      Observations will be available by 10 minutes past the hour.
      By default, the scheduler will check the clock trigger every 10
      seconds, but there is no point in doing this for any interval
      less than around 5 minutes in this case.

      .. hint::

         See :ref:`tutorial-clock-triggers`

      .. spoiler:: Solution warning

         .. note::

            There are a number of ways of writing the dependency.
            You can be correct and not look _exactly_ like this.

         .. code-block:: diff

>            [scheduling]
>                initial cycle point = 20000101T00Z
>            +   [[xtriggers]]
>            +       obs_clock_trigger = wall_clock(PT10M):PT5M
>                [[graph]]
>                    PT3H = """
>            -            get_observations_aldergrove => consolidate_observations
>            +           @obs_clock_trigger => get_observations_aldergrove => consolidate_observations
>            -            get_observations_heathrow => consolidate_observations
>            +           @obs_clock_trigger => get_observations_heathrow => consolidate_observations
>            -            get_observations_aldergrove => consolidate_observations
>            +           @obs_clock_trigger => get_observations_aldergrove => consolidate_observations
>            -            get_observations_camborne => consolidate_observations
>            +           @obs_clock_trigger => get_observations_camborne => consolidate_observations
>                    """
>                    +PT6H/PT6H = """
>                        consolidate_observations => forecast
>                        consolidate_observations[-PT3H] => forecast
>                        consolidate_observations[-PT6H] => forecast
>                        get_rainfall => forecast => post_process_exeter
>                    """

Reviewer: Indentation fix
```suggestion
            [scheduling]
                initial cycle point = 20000101T00Z
           +    [[xtriggers]]
           +        obs_clock_trigger = wall_clock(PT10M):PT5M
                [[graph]]
                    PT3H = """
           -            get_observations_aldergrove => consolidate_observations
           +            @obs_clock_trigger => get_observations_aldergrove => consolidate_observations
           -            get_observations_heathrow => consolidate_observations
           +            @obs_clock_trigger => get_observations_heathrow => consolidate_observations
           -            get_observations_aldergrove => consolidate_observations
           +            @obs_clock_trigger => get_observations_aldergrove => consolidate_observations
           -            get_observations_camborne => consolidate_observations
           +            @obs_clock_trigger => get_observations_camborne => consolidate_observations
                    """
                    +PT6H/PT6H = """
                        consolidate_observations => forecast
                        consolidate_observations[-PT3H] => forecast
                        consolidate_observations[-PT6H] => forecast
                        get_rainfall => forecast => post_process_exeter
                    """
```
